# 高斯玻色采样(Gaussian Boson Sampling)

## 数学背景

<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/gbs.png" width="40%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
高斯玻色采样(GBS)可以作为玻色采样的一种变体， 不同之处在于输入的量子态是高斯压缩态而不是离散的Fock态， 压缩态是高斯态的一种，高斯态指的是这个量子态对应的Wigner函数是高斯分布，比如相干态。 单模压缩态的Wigner函数对应的高斯分布在X，P两个正交分量上会压缩或者拉伸，单模压缩态可以将压缩门作用到真空态上得到，也可以用下面的Fock态基矢展开[2]，需要注意的是这里的Fock态光子数从0到无穷大取偶数， 因此输出的量子态的空间是无限大且只包含偶数光子数的Fock态空间。
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/s.png" width="40%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>

GBS采样概率的理论计算和玻色采样类似， 不同之处在于引入了粒子数分辨探测器和阈值探测器两种情况，对应的计算需要引入hafnian 函数和torotonian函数。
1. 粒子数分辨探测器

在探测端口引入粒子数探测器时对应数学上需要计算hafnian函数，
$2m\times 2m$ 对称矩阵A的hafnian定义如下[3]，
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/haf.png" width="30%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
这里PMP表示所有完美匹配排列的集合，当$n=4$时，$PMP(4) = {(0,1)(2,3),(0,2)(1,3),(0,3)(1,2)}$, $4\times 4$矩阵对应的hafnian如下

$$
haf(B) = B_{0,1}B_{2,3}+B_{0,2}B_{1,3} + B_{0,3}B_{1,2}
$$
在图论中，hafnian计算了图G对应的邻接矩阵A描述的图的完美匹配数(这里图G是无权重，无环的无向图)，比如邻接矩阵$A =\begin{pmatrix}
0&1&1&1\\
1&0&1&1\\
1&1&0&1\\
1&1&1&0
\end{pmatrix}$，haf(A)=3, 对应的完美匹配图如下。
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/f9.png" width="30%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>

当计算的图是二分图时，得到的hanian计算结果就是permanent。
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/per.png" width="40%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
因此任何计算hafnian的算法也可以用来计算permanent, 同样的计算hafnian也是$#P$难问题。

对于粒子数探测器，输出的Fock态$S = (s_1, s_2,..,s_m)$时，对应的$s_i=0,1,2...$,
输出态的概率理论计算如下
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/f5.png" width="30%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
这里$Q,A,X$的定义如下， 
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/f6.png" width="20%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
$Q,A$由输出量子态的协方差矩阵$\Sigma$决定($\Sigma$ 描述的是$a,a^+$的协方差矩阵)，子矩阵$A_s$
由输出的Fock态决定， 具体来说取矩阵$A$的$i, i+m$行和列并且重复$s_i$次来构造$A_s$。如果$s_i=0$，那么就不取对应的行和列,
如果所有的$s_i=1$, 那么对应的子矩阵$A_s = A$。

考虑高斯态是纯态的时候， 矩阵$A$可以写成直和的形式，$A = B \oplus B^*$, $B$是$m\times m$的对称矩阵。这种情况下输出Fock态的概率如下
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/f4.png" width="30%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
这里的子矩阵$B_s$通过取i行和i列并且重复$s_i$次来构造，同时这里hafnian函数计算的矩阵维度减半，可以实现概率计算的加速。

当所有模式输出的光子数$s_i = 0,1$时，对应的$A_s$是A的子矩阵，也对应到邻接矩阵A对应的图G的子图，利用这个性质可以解决很多子图相关的问题，比如稠密子图，
最大团问题等。

2.阈值探测器

使用阈值探测器时对应的输出Fock态概率$S = (s_1, s_2,..,s_m)， s_i\in \{0,1\}$, 此时理论概率的计算需要用到Torontonian函数[4]
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/f7.png" width="20%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/f8.png" width="40%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
这里$O_s = I-(\Sigma^{-1})_s$, 直观上来看对于阈值探测器对应的特定的Fock态输出只需要将多个粒子数分辨探测器对应的Fock态概率求和即可。



# GBS应用到稠密子图问题

1.图和子图

数学上图G的定义如下: 
$$G=(V,E)$$
集合V中的元素称为节点，集合E中的元素时两个节点组成的无序对，称为边。集合V称为点集，E称为边集。 在图的定义中边的概念定义了节点上的一个对称关系，即邻接关系(adjacency relation)。对于两个节点x,y，如果(x,y)是一条边，则称他们是邻接的，因此一张图可以用一个$n\times n$的邻接矩阵A来表示。比如对于四个节点的全连接图对应的A如下
$$A =\begin{pmatrix}
0&1&1&1\\
1&0&1&1\\
1&1&0&1\\
1&1&1&0
\end{pmatrix}$$
haf(A) = 3, 对应的完美匹配数也为3

子图对应的点集和边集分别是图G的点集的子集和边集的子集，稠密子图直观上对应着连接密集的子图，稠密度的定义如下
$$d(G) = \frac{2|E|}{|V|(|V|-1)}$$
$|E|$表示对应的边的条数，$|V|$表示对应的节点个数。那么稠密子图就对应着稠密度很大的子图。

2. GBS采样和稠密子图

参考文献[5]中讨论了图G的完美匹配数和稠密度$d(G)$的关系，hafnian的计算对应图G的完美匹配数，那么hafnian值越大那么图的稠密度越高，由前面讨论可知高斯玻色采样下hafnian值越大也对应着采样概率越高，即概率越高的样本对应的子图的稠密度越大。在使用粒子数分辨探测器时，通过后选择对采样的样本筛选出只有0，1的结果，这些结果中出现概率较高的Fock态映射的子图就对应了稠密子图。同时还可以用经典算法来寻找稠密子图，这里用到的经典算法如下，

a. 选择子图规模大小$[k_{min},k_{max}]$

b.子图规模从大到小搜索(shrinking)，从全图开始，对于$k>k_{min}$, 每次搜索随机删除一个连接数最少的节点，剩下的节点组合成当前规模下的稠密子图。

c.子图规模从小到大搜索(growth)，对于$k<k_{max}$,每次搜索随机添加一个连接数最多的节点，组成当前规模下的稠密子图


## 代码演示

In [19]:
import os
import matplotlib.pyplot as plt
import sys
import torch
dir_ = os.path.abspath('__file__')
sys.path.append(os.path.dirname(os.path.dirname(dir_)))
from data_ import data

In [20]:
import deepquantum as dq
import networkx as nx
from collections import defaultdict
from strawberryfields.apps.subgraph import resize

### 经典算法

这里采用下图中16个节点的图作为例子，这个图可以看作是两部分子图组成，一部分是较为稀疏的子图，对应节点为0到9，另一部分是稠密的子图，
对应的节点为10到15。我们的目标是找到包含6个节点的稠密子图，即[10,11,12,13,14,15]组成的子图。
<div style="margin-right: 15px; border-radius: 10px; background-color: rgb(255， 255， 255); text-align: center;">
    <img src="./fig/graph.png" width="40%"/>
    <p style="padding: 10px; font-size: small; text-align: center; line-height: 0%;">
        <b>
    </p>
</div>
这里的经典算法是基于贪心算法实现的， 即每一次迭代寻找连接数最少的那个节点然后移除就可得到当前规模下的稠密子图， 但是如果有多个节点的连接数相同，那么它会随机选择一个节点移除， 这就导致了目标稠密子图包含的节点有可能被移除，最终导致得到的结果有偏差。

In [21]:
a = data.load_adj('densegraph_adj')
graph = nx.from_numpy_array(a)
s = range(16)
r = resize(s, graph, min_size=1, max_size=15)
r[6], nx.density(graph.subgraph(r[6]))

([5, 10, 11, 13, 14, 15], 0.7333333333333333)

### 量子-经典混合算法

量子-经典混合算法中先通过高斯玻色采样得到概率较高的样本然后转化成对应的子图，这些子图可以作为经典算法的搜索起点，可以有效的提高最后结果的准确度。

这里先load已有的高斯玻色采样数据，采样次数为十万次，gbs.postselect 函数先挑出那些对应子图节点为8,10的样本，然后将这些样本子图作为经典搜索算法的起点，可以得到一个最终收敛到节点为6的子图字典， 字典中包含了节点为6的图密度较大的多个子图， 我们取图密度最大的那个子图就是最终的结果

In [22]:
def search_subgpraph(samples: list,
                    graph: nx.Graph,
                    min_size: int,
                    max_size: int):
    """Get the densest subgraph with size in [min_size, max_size],
        using classical algorithm with samples from GBS
    """
    dic_list = defaultdict(list)
    for i in range(len(samples)):
        temp= samples[i]
        num = 1
        for key in temp.keys():
            if num < 50: # only need 50 samples
                idx = torch.nonzero(torch.tensor(key)).squeeze()
                r = resize(idx.tolist(), graph, min_size=min_size, max_size=max_size)
                for j in range(min_size, max_size+2, 2):
                    density = nx.density(graph.subgraph(r[j]))
                    temp_value = (r[j], np.round(density, 5))
                    if temp_value not in dic_list[j]:
                        dic_list[j].append(temp_value)
                num = num + 1
    return dic_list

In [23]:
# 后处理得到节点为8和10个子图对应的样本
sample_re = data.load_sample('densegraph_sample')
gbs = dq.photonic.GBS_Graph(adj_mat=torch.tensor(a, dtype = torch.float64), cutoff=2)
state = gbs()
subgraph_sample = gbs.postselect(sample_re, [8, 10])

# 采用shrinking 方法得到节点为6 和 8 的稠密子图
dense_sub_graph = search_subgpraph(subgraph_sample, graph, min_size=6, max_size=8)
print(dense_sub_graph[6][0])

([10, 11, 12, 13, 14, 15], 0.93333)


In [ ]:
dense_sub_graph

### 量子算法

量子算法直接将高斯玻色采样后的6个节点对应的样本挑选出来处理，因为根据前面的讨论可以知道，对应的子图越稠密那么其样本出现的概率也就越大。
这里先load已有的高斯玻色采样数据，采样次数为十万次，gbs.postselect 函数先挑出那些对应子图节点为6的样本，然后gbs.graph_density 函数将
这些样本映射成子图再计算子图的图密度，最后按图密度从大到小排列给出对应的子图及其图密度。从最后的结果可以看到，高斯玻色采样成功采到了稠密度最高的6个节点的子图，对应的稠密度为0.9333，对应的节点为[10,11,12,13,14,15]。

In [24]:
sample_re = data.load_sample('densegraph_sample')

In [25]:
subgraph_sample = gbs.postselect(sample_re, [6])
subgraph_density = gbs.graph_density(graph, subgraph_sample[0])
key = list(subgraph_density.keys())
print(key[0],subgraph_density[key[0]])

(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1) [tensor([144]), 0.9333333333333333]


## 附录

[1] Lvovsky, Alexander I. "Squeezed light." Photonics: Scientific Foundations, Technology and Applications 1 (2015): 121-163.

[2]Bromley, Thomas R., et al. "Applications of near-term photonic quantum computers: software and algorithms." Quantum Science and Technology 5.3 (2020): 034010.

[3]Quesada, Nicolás, Juan Miguel Arrazola, and Nathan Killoran. "Gaussian boson sampling using threshold detectors." Physical Review A 98.6 (2018): 062322.

[4]J. M. Arrazola and T. R. Bromley, Physical Review Letters 121, 030503 (2018)